lib/internal/fs/dir.js

```diff
@@ -108,7 +108,10 @@ class Dir {
       throw new ERR_INVALID_CALLBACK(callback);
     }
     const req = new FSReqCallback();
-    req.oncomplete = callback;
+    req.oncomplete = (err) => {
+      this[kDirClosed] = true;
```

Comment:
This means that calling `.close()` twice synchronously can still crash. I would set this flag at the beginning of the function, even if that makes its name somewhat inaccurate.

Reply:
That would mean that if `close()` or `closeSync()` throws an error (for whatever reason), the resource is now unclosable. Maybe that's not a practical problem, though.

Reply:
I think that’s a good thing. Quoting the Linux manpage for `close()`:

> Retrying  the close() after a failure return is the wrong thing to do, since this may cause a reused file descriptor from another thread to be closed.  This can occur because the Linux kernel always releases the file descriptor early in the close operation, freeing it for reuse; the steps that may return an error, such as flushing data to the filesystem or device, occur only later in the close operation.

> Many  other implementations similarly always close the file descriptor (except in the case of EBADF, meaning that the file descriptor was invalid) even if they subsequently report an error on return from close().  POSIX.1 is currently silent on this point, but there are plans to mandate this behavior in the next major release of the standard

tl;dr: If closing failed, you shouldn’t re-try anyway.

Reply:
TIL. 👍